src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs

```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+        {
+            DiagnosticBag bag = DiagnosticBag.GetInstance();
+            try
+            {
+                EnsureSpecialType(SpecialType.System_Int32, bag);
+                EnsureSpecialType(SpecialType.System_IDisposable, bag);
+                EnsureSpecialMember(SpecialMember.System_IDisposable__Dispose, bag);
+
+                // IEnumerator
+                EnsureSpecialType(SpecialType.System_Collections_IEnumerator, bag);
+                EnsureSpecialPropertyGetter(SpecialMember.System_Collections_IEnumerator__Current, bag);
+                EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__MoveNext, bag);
+                EnsureSpecialMember(SpecialMember.System_Collections_IEnumerator__Reset, bag);
+
+                // IEnumerator<T>
+                EnsureSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T, bag);
+                EnsureSpecialPropertyGetter(SpecialMember.System_Collections_Generic_IEnumerator_T__Current, bag);
+
+                if (_isEnumerable)
+                {
+                    // IEnumerable and IEnumerable<T>
+                    EnsureSpecialType(SpecialType.System_Collections_IEnumerable, bag);
+                    EnsureSpecialMember(SpecialMember.System_Collections_IEnumerable__GetEnumerator, bag);
+                    EnsureSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T, bag);
+                    EnsureSpecialMember(SpecialMember.System_Collections_Generic_IEnumerable_T__GetEnumerator, bag);
+                }
+
+                if (bag.HasAnyErrors())
+                {
+                    diagnostics.AddRange(bag);
+                    return false;
+                }
+
+                return true;
+            }
+            finally
+            {
+                bag.Free();
+            }
+        }
+
+        private Symbol EnsureSpecialMember(SpecialMember member, DiagnosticBag bag)
+        {
+            Symbol symbol;
+            Binder.TryGetSpecialTypeMember(F.Compilation, member, body.Syntax, bag, out symbol);
+            return symbol;
+        }
+
+        private void EnsureSpecialType(SpecialType type, DiagnosticBag bag)
+        {
+            Binder.GetSpecialType(F.Compilation, type, body.Syntax, bag);
+        }
+
+        /// <summary>
+        /// Check that the property and its getter exist and collect any use-site errors.
+        /// </summary>
+        private void EnsureSpecialPropertyGetter(SpecialMember member, DiagnosticBag bag)
+        {
+            PropertySymbol symbol = (PropertySymbol)EnsureSpecialMember(member, bag);
+            if (symbol != null)
+            {
+                var getter = symbol.GetMethod;
+                if (getter == null )
+                {
+                    Binder.Error(bag, ErrorCode.ERR_PropertyLacksGet, body.Syntax, symbol);
+                    return;
+                }
+                if (getter.HasUseSiteError)
```

Comment:
Consider getting UseSiteDiagnostic and reporting it if it is not null regardless whether it is an error or not. Warnings are very uncommon, no reason to optimize for them. Also, consider adding an empty line between 'if' statements.